### ISSUE TITLE
Type Narrowing: `if val in Sequence[Literal]` should always narrow to Literals given

### ISSUE THREAD

**Issue by CarliJoy** [python/mypy#16820]
**Feature**

If a value is compared to a sequence of literals, it should be typed narrowed to this Literal.

```python
from typing import reveal_type, Literal, Iterable

def only_ab(val: Literal["a", "b"]):
    ...

def mvp1(strings: Iterable[str]):
    for i in strings:
        if i in ("a", "b"):
            reveal_type(i)  # str instead of Literal["a", "b"]
```

Set of multiple examples [in MyPy Play](https://mypy-play.net/?mypy=master&python=3.12&gist=51c2c1423471c3f8e574705da5c52317)

**Pitch**

See https://github.com/python/mypy/issues/16813 for an example why this is useful.
In general working with TypedDicts in for loop this helps a lot reducing duplicate code / keys.

#16819 is probably required to in order for this to work.

**Comment by JelleZijlstra** [comment]
As with #16819, I think there's some small risk of new false positives, but on balance this proposed behavior is likely better.

**Comment by CarliJoy** [comment]
As written in #16819, this already works in [pyright](https://pyright-play.net/?pythonVersion=3.12&code=GYJw9gtgBALgngBwJYDsDmUkQWEMogCmAboQIYA2A%2BvAoQDRQAySMhIljAkmxwEYVCAWABQogCaFgUMCgpwqZPgApilAFzNW7SgG0ARGX2N9ffQF0AlOtFQ7UAHRPRLkZOkRiCAIzKAzjAgqGh%2Bmjw6AoS6ASBWNiL2UMC4mJgoUDHBobaJ9kjSSGlQyobGUKb61jm5uUSklDSIhMpIlnYAxBmB1TX2svKKKq1QnVCAMuRdIFAoYPhavHqlJmbmPYn5qQC8m%2BVG8b25-QpKLW2jEzHTs3YsCxQGRssWru5QnggATP6BWWELkdFAnEemoKKFYABXBCCXS3HT3UrmRhwjgIlbmKA7EqPcpmSw9ZJTQqoSa-NZ5ApFUHZBIHex1cjUWjNYYjOwXQJXOYoxY4iqrWl0mRyY5DfGC3IbQrbXb6fZC4UDE6s86TLk3bSoh5lfnkuwSKRvLwAZm%2BQXQ4PC-BhMWBEupmgoSACsM1vJ16MxUG1TwFiUJqRJmQt8pqUqplBpCoIJEZjTopzZMT1hxFg0TqsuM25boRfJWKfDMtKobpR3TKvZauzGruPtxzwlBxTiQNHi8ABYzb8oFalDCeXmPTqAMYWO3%2BlLE9LBkKl9aUknY4dVJu9BkNZkZyYtvpp5VtNnjavXebw%2Bu6td2Is7Eu7uzlg8dKAAd1wAGs-C9De8AKzdkNe3%2BAdcwvMwTDHcwJ3sB1IWhKJB21JEzy1fkMSxJYG3FSciSKWco16cMSQde8Y3qJkmm3ZMr1TJUxWfY8s1PRDMMvA4b1ledekfei2TfEBPxTNsjQQAA2AC5z7AEWPzUdx1XRJYKdF0ZOHKCvTAxscMDGcfhDQtF3SEiaMSDcKITYZOmo6MeMTI8OSmGsUPdX0DK2W89lI2zK34z9WxM%2BxSOE3y-GUJQR00VSnggiwFPsMhItAxEoAAHygAA5WRCC9TKUGECUNnCzEPLlFMyC9cLv2kEKqGzMKUDgWdNBiOK7AS5yhwxNLcuynYep6QqGtnYrOLKiqhr0jA2UIAAPBAiD8PwkFkKAAAsyD8WAmh3EQgA) as expected

**Comment by asottile-sentry** [comment]
seems to be a duplicate of #12535

**Comment by JelleZijlstra** [comment]
Thanks, agree. I feel there's a lot of duplicates among this type of issue ("mypy doesn't understand some type narrowing idiom") but I haven't spent the time to sort them out.

Duplicate of #12535
